### ISSUE TITLE
Improve cache_labels() to tolerate out of bound labels.

### ISSUE THREAD

**Issue by pocketvision** [ultralytics/yolov5#2785]
## 🚀 Feature
Improve cache_labels() to tolerate out of bound labels. Or at least, ignore the specific labels whose iou with the image frame is lower than a threshold, other than filter out the total sample.

## Motivation

1、We may have normal out of bound labels, which should not be filtered out.
![label_out_of_bound](https://user-images.githubusercontent.com/20749514/114650353-b1a1a680-9d14-11eb-8fa6-4b3aeb972f2a.jpg)
2、We may filter out a lot of samples because of some specific out of bound labels.
![one_label_out_of_bound](https://user-images.githubusercontent.com/20749514/114651117-eeba6880-9d15-11eb-9227-941420050855.jpg)

**Comment by glenn-jocher** [comment]
@daikankan all label bounding box values must be 0.0 <= x <= 1.0. These label checks are a feature to visibly announce label errors and reject these labels from entering the train/va/test datasets used by YOLOv5. Accepting out of bound labels or making assumptions about their errors and applying our own modifications to your labels is probably not a good idea as this will lead to silent errors.

**Comment by daikankan** [comment]
@glenn-jocher What's the silent errors? Cause actually，I do found it's hard for training when I comment these two lines:
                        # assert (l >= 0).all(), 'negative labels'
                        # assert (l[:, 1:] <= 1).all(), 'non-normalized or out of bounds coordinate labels'
convergence hard:
![results](https://user-images.githubusercontent.com/20749514/115180162-87caf400-a107-11eb-80ad-9ac399d71d90.png)

![labels](https://user-images.githubusercontent.com/20749514/115179943-10956000-a107-11eb-9de7-29605e890754.jpg)
![labels_correlogram](https://user-images.githubusercontent.com/20749514/115179946-12f7ba00-a107-11eb-9d0c-9bccec92ee66.jpg)
![train_batch2](https://user-images.githubusercontent.com/20749514/115179973-2014a900-a107-11eb-9ec1-e0fe9b7f6663.jpg)
But the sample plot seems currect, I'm training face detection for 4 orients support.

**Comment by glenn-jocher** [comment]
👋 Hello! Thanks for asking about improving training results. Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings before considering any changes**. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥1.5k images per class
- **Instances per class.** ≥10k instances (labeled objects) per class total
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total).

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x will produce better results in nearly all cases, but have more parameters and are slower to run. For mobile applications we recommend YOLOv5s/m, for cloud or desktop applications we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

To start training from pretrained weights simply pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
```

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://user-images.githubusercontent.com/26833433/103595982-ab986000-4eb1-11eb-8c57-4726261b0a88.png"></p>

## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
http://karpathy.github.io/2019/04/25/recipe/

**Comment by pocketvision** [comment]
Thanks

**Comment by glenn-jocher** [comment]
@daikankan you want to train with all default settings before considering any modifications, otherwise you can't compare to the existing results, and you don't have a baseline to start improving on.

**Comment by daikankan** [comment]
hyp['box'] *= 3. / nl  # scale to layers
    hyp['cls'] *= nc / 80. * 3. / nl  # scale to classes and layers
    hyp['obj'] *= (imgsz / 640) ** 2 * 3. / nl  # scale to image size and layers

The default hyp should be the reason, and I'll make a search for the hyp parameters for my own task. 

Thanks again for the yolov5 framework @glenn-jocher , based on which I made some interesting tasks, such as  instance segmentation ^ ^:
![000000473974](https://user-images.githubusercontent.com/20749514/115235795-29286900-a14d-11eb-93b3-7109e849f7e6.jpg)![000000473974](https://user-images.githubusercontent.com/20749514/115235805-2f1e4a00-a14d-11eb-97a8-772406b77336.jpg)
![3](https://user-images.githubusercontent.com/20749514/115236092-88867900-a14d-11eb-95c9-d78dbc31ee5c.jpg)![3](https://user-images.githubusercontent.com/20749514/115236111-8de3c380-a14d-11eb-82d2-32c2cbd7c350.jpg)
